import "combat.proto";

package protocol;

// 消息类型
enum MessageType {
  MESSAGE_LOGIN_REQUEST = 1;                                        // 登陆请求
  MESSAGE_LOGIN_RESPONSE = 2;                                       // 登陆回应

  MESSAGE_COMBAT_REQUEST = 3;                                       // 战斗请求
  MESSAGE_COMBAT_RESPONSE = 4;                                      // 战斗回应

  MESSAGE_COMBAT_BEGIN_SYNC = 5;                                    // 战斗开始同步
  MESSAGE_COMBAT_SETTLEMENT_SYNC = 6;                               // 战斗结算同步

  MESSAGE_COMBAT_ACTION_REQUEST = 1001;                             // 行动请求
  MESSAGE_COMBAT_ACTION_RESPONSE = 1002;                            // 行动回应
  MESSAGE_COMBAT_ACTION_SYNC = 1003;                                // 行动同步
};

// ----------------- message body ------------------

// MESSAGE_LOGIN_REQUEST
message MessageLoginRequest {
  required string name = 1;                                         // 名称
}

// MESSAGE_LOGIN_RESPONSE
message MessageLoginResponse {
  required bool result = 1;                                         // 结果
}

// MESSAGE_COMBAT_REQUEST
message MessageCombatRequest {
  repeated int32 warrior_id = 1;                                    // 出战士兵
}

// MESSAGE_COMBAT_RESPONSE
message MessageCombatResponse {
  required bool result = 1;                                         // 结果
  optional uint64 host_id = 2;                                      // 当前角色的代理id
  optional int32 camp_id = 3;                                       // 阵营
  optional .protocol.CombatDescription combat_description = 4;      // 战斗的基本描述
  optional .protocol.CombatStatusImage status_image = 5;            // 状态镜像
  repeated int32 combat_warrior_id = 6;                             // 出战斗士兵列表
}

// MESSAGE_COMBAT_BEGIN_SYNC
message MessageCombatBeginSync {
  required uint64 host_id = 1;                                      // 当前角色的代理id
  required int32 camp_id = 2;                                       // 当前角色的阵营id
  required .protocol.CombatDescription combat_description = 3;      // 战斗的基本描述
  required .protocol.CombatStatusImage init_status_image = 4;       // 战斗初始化数据
  repeated int32 combat_warrior_id = 5;                             // 出战斗士兵列表
}

// MESSAGE_COMBAT_SETTLEMENT_SYNC
message MessageCombatSettlementSync {
  required .protocol.CombatSettlement settlement = 1;               // 结算数据
}

// MESSAGE_COMBAT_ACTION_REQUEST
message MessageCombatActionRequest {
  required .protocol.CombatAction action = 1;                       // 动作
}

// MESSAGE_COMBAT_ACTION_RESPONSE
message MessageCombatActionResponse {
  required bool result = 1;                                         // 结果
}

// MESSAGE_COMBAT_ACTION_SYNC
message MessageCombatActionSync {
  required .protocol.CombatAction action = 1;                       // 动作
}
